### ISSUE TITLE
Data from box opens is not zeroed on drop

### ISSUE THREAD

**Issue by jmacmahon** [sodiumoxide/sodiumoxide#417]
[box_::open](https://docs.rs/sodiumoxide/0.2.5/sodiumoxide/crypto/box_/curve25519xsalsa20poly1305/fn.open.html) and [sealedbox::open](https://docs.rs/sodiumoxide/0.2.5/sodiumoxide/crypto/sealedbox/curve25519blake2bxsalsa20poly1305/fn.open.html) return a `Vec<u8>` on success, which is not a newtype like `Key` etc., and therefore does not have the protection of being zeroed when being dropped (e.g. for `Key` we have [this](https://docs.rs/sodiumoxide/0.2.5/src/sodiumoxide/newtype_macros.rs.html#216-221) `Drop` implementation)

Would it be a good idea to add a newtype for decrypted data which implements this `Drop` safety feature?

Alternatively we could mimic the C API by providing a buffer to be filled, with the consumer ensuring memory protection, like in [the pwhash example](https://docs.rs/sodiumoxide/0.2.5/sodiumoxide/crypto/pwhash/index.html#example-key-derivation).

**Comment by mso42** [comment]
Some things to consider:

Although we initialise the Vec with `Vec::with_capacity`, thus preventing reallocation, it is possible that the consumer may append or generally mutate the `Vec` in such a way that it gets reallocated. It would therefore not be possible to zero the previously-allocated buffer. [zeroize](https://github.com/iqlusioninc/crates/tree/develop/zeroize) calls this out explicitly in its [`impl Zeroize for Vec<Z>`](https://docs.rs/zeroize/1.2.0/zeroize/trait.Zeroize.html#method.zeroize).

One solution to this could be to return some kind of newtype wrapping a `Box<[u8]>` instance, implementing secure `Drop` on that.

However in `no-std` we don't have `Box`. But then, we don't have `Vec` either... so how do these functions even work in those environments? Accepting a `&mut [u8]` and returning `()` would push both of these issues out of scope and up to the consumer.

So the way I see it, we have two options:
- Wrap up a `Box<[u8]>` in a newtype, implementing a secure `Drop` ourselves.
- Accept a `&mut [u8]`, leaving secure drop (and allocation in general) up to the consumer.

Thoughts?

**Comment by mso42** [comment]
Oh, but if we accept `&mut [u8]`, we then have to deal with the possiblity that the buffer may be too short for the decrypted message.

**Comment by adamreichold** [comment]
In all discussions of zeroing-on-drop, please keep in mind that this is always best effort only as one generally does not have the necessary control over contents of physical memory in virtual memory systems. IMHO, this suggest carefully weighing the ergonomic and performance implications against the expected benefits.

Please also note that the plain text and the key material usually have different threat scenarios: If I am able to look at the plain text once, I have circumvented encryption once. If I have copied the secret key, I have have circumvented every future encryption using that key. This would at least suggest that using a zeroing newtype for keys but not for plain text is a reasonable trade-off.

**Comment by mso42** [comment]
Good points, all.

Personally I think we could make it not too un-ergonomic with a wrapped Box<[u8]>, by implementing `AsRef<[u8]>` and `From<_> for Vec<u8>` and other convenience traits.

**Comment by adamreichold** [comment]
> Personally I think we could make it not too un-ergonomic

I tend to agree that a newtype wrapper could be workable. Considering that in contrast to keys, the plain text can potentially be a large amount of data, I am more concerned about performance issues, e.g. the cost of zeroing large amounts of data could be to large to impose upon all users of the library.

**Comment by mso42** [comment]
Ah, true. I hadn't thought of that.

I wonder if returning a plain `Box<[u8]>` (i.e. not a newtype) would be an acceptable solution -- that would prevent the re-allocation problem if the user did choose to zero the data, but not impede performance for large buffers.

**Comment by dnaq** [comment]
TBH I think those apis should be deprecated since libsodium has supported
detached authentication tags for a long time now.

**Comment by mso42** [comment]
> TBH I think those apis should be deprecated since libsodium has supported detached authentication tags for a long time now.

You mean that the `open_detached` functions mutate the ciphertext buffer and therefore this is not a problem?

**Comment by mso42** [comment]
Is the idea that the combined functions in libsodium will eventually be deprecated? If so, I see your point.

But otherwise, shouldn't we try to support as much of libsodium as possible, and that includes the combined functions?

**Comment by dnaq** [comment]
The combined functions are basically the same as the authentication tag
prepended to the ciphertext, so if needed for some reason they’re trivially
implementable by the user.

But for most cases like when sending packets over the network or writing
ciphertext to files it’s more efficient to not have to allocate before
sending. It will at least lower memory pressure, and it works in no_std
environments.

**Comment by dnaq** [comment]
Closing due to deprecation of the library.
